pwndbg/gdblib/arch.py

```diff
@@ -66,12 +73,154 @@ def get_thumb_mode_string() -> Literal["arm", "thumb"] | None:
     return None if thumb_bit is None else "thumb" if thumb_bit == 1 else "arm"
 
 
-arch = Arch("i386", typeinfo.ptrsize, "little")
+# def update_arch_hook(arch: Architecture):
+#     # This might be due to starting a new binary or connecting to another remote process.
+#     # This function will will be pass ed the architecture
+#     # The specific debugger (GDB/LLDB) will detect the arch
+#     # and call this function appropriately
+#     arch = architecture
 
-name: str
-ptrsize: int
-ptrmask: int
-endian: Literal["little", "big"]
+ArchType = Literal[
+    "i386",
+    "x86-64",
+    "rv32",
+    "rv64",
+    "mips",
+    "sparc",
+    "arm",
+    "iwmmxt",
+    "armcm",
+    "aarch64",
+    "powerpc",
+]
+EndianType = Literal["little", "big"]
+
+FMT_LITTLE_ENDIAN = {1: "B", 2: "<H", 4: "<I", 8: "<Q"}
+FMT_BIG_ENDIAN = {1: "B", 2: ">H", 4: ">I", 8: ">Q"}
+
+
+class PwndbgArchitecture:
+    # Registry of all instances
+    arch_registry: Dict[ArchType, PwndbgArchitecture] = {}
+
+    @staticmethod
+    def get_arch(name: ArchType) -> PwndbgArchitecture:
+        # If a custom class has not been registered for the architecture, use base implementation
+        if name not in PwndbgArchitecture.arch_registry:
+            PwndbgArchitecture.arch_registry[name] = PwndbgArchitecture(name)
+
+        return PwndbgArchitecture.arch_registry[name]
+
+    def __init__(self, name: ArchType) -> None:
+        """
+        Calling the constructor will register the class with global list of PwndbgArchitectures
+        """
+        if name is not None:
+            self.arch_registry[name] = self
+
+        self.name: ArchType = name
```

Comment:
Can we add a `__slots__` and move all fields of such class there?

I belivee that this object should never get new fields attached to it, so it probably makes sense to use `__slots__` here, especially that it can also be accessed many times.